src/server/game/Battlegrounds/BattlegroundQueue.cpp

```diff
@@ -424,35 +424,60 @@ void BattlegroundQueue::FillPlayersToBG(Battleground* bg, const int32 aliFree, c
+        case BG_QUEUE_INVITATION_TYPE_NO_BALANCE:
+            return;
+        case BG_QUEUE_INVITATION_TYPE_BALANCED:
+            invDiff = 1;
+        case BG_QUEUE_INVITATION_TYPE_EVEN:
+            invDiff = 0;
+        default:
+            return;
         }
-        // if results in more horde players than alliance:
-        else
+
+        // balance the teams based on the difference allowed
+        while (abs(aliDiff - hordeDiff) > invDiff && (m_SelectionPools[TEAM_HORDE].GetPlayerCount() > 0 || m_SelectionPools[TEAM_ALLIANCE].GetPlayerCount() > 0))
         {
-            // no more horde in pool, invite whatever we can from alliance
-            if (!m_SelectionPools[TEAM_HORDE].GetPlayerCount())
-                break;
+            // if results in more alliance players than horde:
+            if (aliDiff < hordeDiff)
+            {
+                // no more alliance in pool, invite whatever we can from horde
+                if (!m_SelectionPools[TEAM_ALLIANCE].GetPlayerCount())
+                    break;
 
-            // kick horde, returns true if kicked more than needed, so then try to fill up
-            if (m_SelectionPools[TEAM_HORDE].KickGroup(aliDiff - hordeDiff))
-                for (; Horde_itr != m_QueuedGroups[bracket_id][BG_QUEUE_NORMAL_HORDE].end() && m_SelectionPools[TEAM_HORDE].AddGroup((*Horde_itr), hordeFree >= aliDiff ? hordeFree - aliDiff : 0); ++Horde_itr);
-        }
+                // kick alliance, returns true if kicked more than needed, so then try to fill up
+                if (m_SelectionPools[TEAM_ALLIANCE].KickGroup(hordeDiff - aliDiff))
+                    for (; Ali_itr != m_QueuedGroups[bracket_id][BG_QUEUE_NORMAL_ALLIANCE].end() && m_SelectionPools[TEAM_ALLIANCE].AddGroup((*Ali_itr), aliFree >= hordeDiff ? aliFree - hordeDiff : 0); ++Ali_itr);
+            }
+            // if results in more horde players than alliance:
+            else
+            {
+                // no more horde in pool, invite whatever we can from alliance
+                if (!m_SelectionPools[TEAM_HORDE].GetPlayerCount())
+                    break;
+
+                // kick horde, returns true if kicked more than needed, so then try to fill up
+                if (m_SelectionPools[TEAM_HORDE].KickGroup(aliDiff - hordeDiff))
+                    for (; Horde_itr != m_QueuedGroups[bracket_id][BG_QUEUE_NORMAL_HORDE].end() && m_SelectionPools[TEAM_HORDE].AddGroup((*Horde_itr), hordeFree >= aliDiff ? hordeFree - aliDiff : 0); ++Horde_itr);
+            }
 
-        // recalculate free space after adding
-        aliDiff = aliFree - int32(m_SelectionPools[TEAM_ALLIANCE].GetPlayerCount());
-        hordeDiff = hordeFree - int32(m_SelectionPools[TEAM_HORDE].GetPlayerCount());
+            // recalculate free space after adding
+            aliDiff = aliFree - int32(m_SelectionPools[TEAM_ALLIANCE].GetPlayerCount());
```

Comment:
C style casts should be avoided. I would probably look into just using a uint32 earlier on as aliDiff/aliFree. Same for Horde. Use static cast if you can't change them to uint32